File: sdk/go/client/account.go
package client

import (
	"context"

	clientv1 "github.com/torchwooddev/torchwood/genproto/client/v1"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// AccountService 封装 Client API 的 Account 服务。
type AccountService struct{ c *Client }

// SignUp 使用邮箱/密码注册新账户；成功（非 MFA 分支）后自动保存 token。
func (a *AccountService) SignUp(ctx context.Context, email, password, name string) (*clientv1.SignUpResponse, error) {
	resp, err := a.c.account.SignUp(ctx, &clientv1.SignUpRequest{
		Email:     email,
		Password:  password,
		Name:      name,
		ProjectId: a.c.cfg.ProjectID,
	})
	if err != nil {
		return nil, err
	}
	if !resp.MfaRequired && resp.Tokens != nil && resp.Tokens.AccessToken != "" {
		if err := a.c.saveTokens(resp.Tokens); err != nil {
			return nil, err
		}
	}
	return resp, nil
}

// SignIn 使用邮箱/密码登录；成功（非 MFA 分支）后自动保存 token。
func (a *AccountService) SignIn(ctx context.Context, email, password string) (*clientv1.SignInResponse, error) {
	resp, err := a.c.account.SignIn(ctx, &clientv1.SignInRequest{
		Email:     email,
		Password:  password,
		ProjectId: a.c.cfg.ProjectID,
	})
	if err != nil {
		return nil, err
	}
	if !resp.MfaRequired && resp.Tokens != nil && resp.Tokens.AccessToken != "" {
		if err := a.c.saveTokens(resp.Tokens); err != nil {
			return nil, err
		}
	}
	return resp, nil
}

// RefreshToken 用刷新令牌换取新令牌并保存。
func (a *AccountService) RefreshToken(ctx context.Context, refreshToken string) (*clientv1.RefreshTokenResponse, error) {
	resp, err := a.c.account.RefreshToken(ctx, &clientv1.RefreshTokenRequest{
		ProjectId:    a.c.cfg.ProjectID,
		RefreshToken: refreshToken,
	})
	if err != nil {
		return nil, err
	}
	if resp.Tokens != nil && resp.Tokens.AccessToken != "" {
		if err := a.c.saveTokens(resp.Tokens); err != nil {
			return nil, err
		}
	}
	return resp, nil
}

// Me 返回当前登录账户信息。
func (a *AccountService) Me(ctx context.Context) (*clientv1.Account, error) {
	return a.c.account.Me(ctx, &clientv1.MeRequest{ProjectId: a.c.cfg.ProjectID})
}

// SignOut 注销当前会话；成功或 token 已失效（Unauthenticated）都清空本地 token。
func (a *AccountService) SignOut(ctx context.Context) error {
	_, err := a.c.account.SignOut(ctx, &clientv1.SignOutRequest{ProjectId: a.c.cfg.ProjectID})
	if err == nil || status.Code(err) == codes.Unauthenticated {
		a.c.clearTokens()
	}
	return err
}

// UpdateAccount 更新账户资料；name/email 传 nil 表示不修改，传指针（含空串）表示更新/清空。
func (a *AccountService) UpdateAccount(ctx context.Context, name, email *string, password, oldPassword string) (*clientv1.Account, error) {
	return a.c.account.UpdateAccount(ctx, &clientv1.UpdateAccountRequest{
		Name:        name,
		Email:       email,
		Password:    password,
		OldPassword: oldPassword,
	})
}

// ListSessions 列出当前账户的全部会话（Session.current 标记当前会话）。
func (a *AccountService) ListSessions(ctx context.Context) (*clientv1.ListSessionsResponse, error) {
	return a.c.account.ListSessions(ctx, &clientv1.ListSessionsRequest{})
}

// DeleteSession 删除指定会话（删除当前会话即登出）。
func (a *AccountService) DeleteSession(ctx context.Context, sessionID string) error {
	_, err := a.c.account.DeleteSession(ctx, &clientv1.DeleteSessionRequest{SessionId: sessionID})
	return err
}

// DeleteSessions 删除全部会话；keepCurrent 为 true 时保留当前会话。
func (a *AccountService) DeleteSessions(ctx context.Context, keepCurrent bool) error {
	_, err := a.c.account.DeleteSessions(ctx, &clientv1.DeleteSessionsRequest{KeepCurrent: keepCurrent})
	return err
}

// GetPrefs 读取用户偏好（JSON）。
func (a *AccountService) GetPrefs(ctx context.Context) (*clientv1.GetPrefsResponse, error) {
	return a.c.account.GetPrefs(ctx, &clientv1.GetPrefsRequest{})
}

// UpdatePrefs 全量替换用户偏好（JSON）。
func (a *AccountService) UpdatePrefs(ctx context.Context, prefs map[string]any) (*clientv1.GetPrefsResponse, error) {
	st, err := toStruct(prefs)
	if err != nil {
		return nil, err
	}
	return a.c.account.UpdatePrefs(ctx, &clientv1.UpdatePrefsRequest{Prefs: st})
}

// CreateEmailOTP 发送邮箱验证码登录请求，返回 challenge_id。
func (a *AccountService) CreateEmailOTP(ctx context.Context, email string) (*clientv1.ChallengeResponse, error) {
	return a.c.account.CreateEmailOTP(ctx, &clientv1.CreateEmailOTPRequest{
		ProjectId: a.c.cfg.ProjectID,
		Email:     email,
	})
}

// CreateEmailOTPSession 用邮箱验证码完成登录；成功后自动保存 token。
func (a *AccountService) CreateEmailOTPSession(ctx context.Context, email, challengeID, otp string) (*clientv1.SignInResponse, error) {
	resp, err := a.c.account.CreateEmailOTPSession(ctx, &clientv1.CreateEmailOTPSessionRequest{
		ProjectId:   a.c.cfg.ProjectID,
		Email:       email,
		ChallengeId: challengeID,
		Otp:         otp,
	})
	if err != nil {
		return nil, err
	}
	if err := a.saveSignInTokens(resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// CreateOAuth2Session 获取 OAuth2 授权跳转地址（匿名）。
func (a *AccountService) CreateOAuth2Session(ctx context.Context, provider, success, failure string) (*clientv1.CreateOAuth2SessionResponse, error) {
	return a.c.account.CreateOAuth2Session(ctx, &clientv1.CreateOAuth2SessionRequest{
		Provider:  provider,
		ProjectId: a.c.cfg.ProjectID,
		Success:   success,
		Failure:   failure,
	})
}

// CreateOAuth2TokenSession 用 OAuth2 回调 code 换取登录态；成功后自动保存 token。
func (a *AccountService) CreateOAuth2TokenSession(ctx context.Context, provider, success, failure, code, state string) (*clientv1.SignInResponse, error) {
	resp, err := a.c.account.CreateOAuth2TokenSession(ctx, &clientv1.CreateOAuth2TokenSessionRequest{
		Provider:  provider,
		ProjectId: a.c.cfg.ProjectID,
		Success:   success,
		Failure:   failure,
		Code:      code,
		State:     state,
	})
	if err != nil {
		return nil, err
	}
	if err := a.saveSignInTokens(resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// CreatePhoneOTP 发送短信验证码登录请求，返回 challenge_id。
func (a *AccountService) CreatePhoneOTP(ctx context.Context, phone string) (*clientv1.ChallengeResponse, error) {
	return a.c.account.CreatePhoneOTP(ctx, &clientv1.CreatePhoneOTPRequest{
		ProjectId: a.c.cfg.ProjectID,
		Phone:     phone,
	})
}

// CreatePhoneOTPSession 用短信验证码完成登录；成功后自动保存 token。
func (a *AccountService) CreatePhoneOTPSession(ctx context.Context, phone, challengeID, otp string) (*clientv1.SignInResponse, error) {
	resp, err := a.c.account.CreatePhoneOTPSession(ctx, &clientv1.CreatePhoneOTPSessionRequest{
		ProjectId:   a.c.cfg.ProjectID,
		Phone:       phone,
		ChallengeId: challengeID,
		Otp:         otp,
	})
	if err != nil {
		return nil, err
	}
	if err := a.saveSignInTokens(resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// CreateWeChatMiniProgramSession 用微信小程序 code 完成登录；成功后自动保存 token。
func (a *AccountService) CreateWeChatMiniProgramSession(ctx context.Context, code string) (*clientv1.SignInResponse, error) {
	resp, err := a.c.account.CreateWeChatMiniProgramSession(ctx, &clientv1.CreateWeChatMiniProgramSessionRequest{
		ProjectId: a.c.cfg.ProjectID,
		Code:      code,
	})
	if err != nil {
		return nil, err
	}
	if err := a.saveSignInTokens(resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// CreateAnonymousSession 创建匿名会话；成功后自动保存 token。
func (a *AccountService) CreateAnonymousSession(ctx context.Context) (*clientv1.SignInResponse, error) {
	resp, err := a.c.account.CreateAnonymousSession(ctx, &clientv1.CreateAnonymousSessionRequest{
		ProjectId: a.c.cfg.ProjectID,
	})
	if err != nil {
		return nil, err
	}
	if err := a.saveSignInTokens(resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// CreateOAuth2LinkSession 获取第三方账号绑定授权跳转地址（需已登录）。
func (a *AccountService) CreateOAuth2LinkSession(ctx context.Context, provider, success, failure string) (*clientv1.CreateOAuth2SessionResponse, error) {
	return a.c.account.CreateOAuth2LinkSession(ctx, &clientv1.CreateOAuth2LinkSessionRequest{
		Provider:  provider,
		ProjectId: a.c.cfg.ProjectID,
		Success:   success,
		Failure:   failure,
	})
}

// CreateOAuth2LinkTokenSession 用 OAuth2 回调 code 绑定第三方账号（需已登录）。
func (a *AccountService) CreateOAuth2LinkTokenSession(ctx context.Context, provider, code, state string) (*clientv1.Account, error) {
	return a.c.account.CreateOAuth2LinkTokenSession(ctx, &clientv1.CreateOAuth2LinkTokenSessionRequest{
		Provider:  provider,
		ProjectId: a.c.cfg.ProjectID,
		Code:      code,
		State:     state,
	})
}

// CreateVerification 发送邮箱验证邮件（url 为带 {{code}} 占位的确认链接模板）。
func (a *AccountService) CreateVerification(ctx context.Context, url string) (*clientv1.CreateVerificationResponse, error) {
	return a.c.account.CreateVerification(ctx, &clientv1.CreateVerificationRequest{
		ProjectId: a.c.cfg.ProjectID,
		Url:       url,
	})
}

// UpdateVerification 用邮件中的 secret 确认邮箱（公开方法，无需登录）。
func (a *AccountService) UpdateVerification(ctx context.Context, userID, secret string) (*clientv1.Account, error) {
	return a.c.account.UpdateVerification(ctx, &clientv1.UpdateVerificationRequest{
		ProjectId: a.c.cfg.ProjectID,
		UserId:    userID,
		Secret:    secret,
	})
}

// CreateRecovery 发送密码找回邮件（公开方法）。
func (a *AccountService) CreateRecovery(ctx context.Context, email, url string) error {
	_, err := a.c.account.CreateRecovery(ctx, &clientv1.CreateRecoveryRequest{
		ProjectId: a.c.cfg.ProjectID,
		Email:     email,
		Url:       url,
	})
	return err
}

// UpdateRecovery 用邮件中的 secret 重置密码（公开方法）。
func (a *AccountService) UpdateRecovery(ctx context.Context, userID, secret, password string) error {
	_, err := a.c.account.UpdateRecovery(ctx, &clientv1.UpdateRecoveryRequest{
		ProjectId: a.c.cfg.ProjectID,
		UserId:    userID,
		Secret:    secret,
		Password:  password,
	})
	return err
}

// ListFactors 列出 MFA 因子。
func (a *AccountService) ListFactors(ctx context.Context) (*clientv1.ListFactorsResponse, error) {
	return a.c.account.ListFactors(ctx, &clientv1.ListFactorsRequest{})
}

// CreateTOTPFactor 创建 TOTP 因子；secret/otpauth_url 仅本次响应返回明文。
func (a *AccountService) CreateTOTPFactor(ctx context.Context) (*clientv1.TOTPFactor, error) {
	return a.c.account.CreateTOTPFactor(ctx, &clientv1.CreateTOTPFactorRequest{})
}

// VerifyTOTPFactor 校验 code 并激活 TOTP 因子。
func (a *AccountService) VerifyTOTPFactor(ctx context.Context, factorID, code string) (*clientv1.Factor, error) {
	return a.c.account.VerifyTOTPFactor(ctx, &clientv1.VerifyTOTPFactorRequest{
		FactorId: factorID,
		Code:     code,
	})
}

// DeleteFactor 删除 MFA 因子；verified 因子需携带 TOTP 二次验证 code，
// pending 因子可传空串（R05-P1-4）。
func (a *AccountService) DeleteFactor(ctx context.Context, factorID, code string) error {
	_, err := a.c.account.DeleteFactor(ctx, &clientv1.DeleteFactorRequest{
		FactorId: factorID,
		Code:     code,
	})
	return err
}

// CreateMFASession 登录二次验证：携带 signIn/signUp 返回的 challenge_token 完成挑战；
// 成功后自动保存 token。
func (a *AccountService) CreateMFASession(ctx context.Context, challengeToken, factorID, code string) (*clientv1.SignInResponse, error) {
	resp, err := a.c.account.CreateMFASession(ctx, &clientv1.CreateMFASessionRequest{
		ProjectId:      a.c.cfg.ProjectID,
		ChallengeToken: challengeToken,
		FactorId:       factorID,
		Code:           code,
	})
	if err != nil {
		return nil, err
	}
	if err := a.saveSignInTokens(resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// CreateJWT 用当前会话换取一次性 JWT（用于服务端安全回调/Webhook）。
func (a *AccountService) CreateJWT(ctx context.Context) (*clientv1.CreateJWTResponse, error) {
	return a.c.account.CreateJWT(ctx, &clientv1.CreateJWTRequest{})
}

// CreateMagicURLSession 发送 Magic URL 登录邮件（公开方法）。
func (a *AccountService) CreateMagicURLSession(ctx context.Context, email, url string) (*clientv1.ChallengeResponse, error) {
	return a.c.account.CreateMagicURLSession(ctx, &clientv1.CreateMagicURLSessionRequest{
		ProjectId: a.c.cfg.ProjectID,
		Email:     email,
		Url:       url,
	})
}

// UpdateMagicURLSession 用邮件中的 secret 完成 Magic URL 登录；成功后自动保存 token。
func (a *AccountService) UpdateMagicURLSession(ctx context.Context, userID, secret string) (*clientv1.SignInResponse, error) {
	resp, err := a.c.account.UpdateMagicURLSession(ctx, &clientv1.UpdateMagicURLSessionRequest{
		ProjectId: a.c.cfg.ProjectID,
		UserId:    userID,
		Secret:    secret,
	})
	if err != nil {
		return nil, err
	}
	if err := a.saveSignInTokens(resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// ListLogs 返回最近 limit 条账号操作日志。
func (a *AccountService) ListLogs(ctx context.Context, limit int32) (*clientv1.ListLogsResponse, error) {
	return a.c.account.ListLogs(ctx, &clientv1.ListLogsRequest{Limit: limit})
}

// saveSignInTokens 非 MFA 且返回 token 时自动保存（SignIn 系列响应统一处理）。
func (a *AccountService) saveSignInTokens(resp *clientv1.SignInResponse) error {
	if resp == nil || resp.MfaRequired || resp.Tokens == nil || resp.Tokens.AccessToken == "" {
		return nil
	}
	return a.c.saveTokens(resp.Tokens)
}

File: sdk/typescript/src/client/account.ts
import type { HttpTransport } from "../http.js";
import type {
  Account,
  AuthResult,
  Factor,
  LogEntry,
  Session,
  TokenBundle,
  TOTPFactor,
} from "../types.js";

export class AccountService {
  constructor(private readonly http: HttpTransport) {}

  async signUp(input: {
    email: string;
    password: string;
    name: string;
  }): Promise<AuthResult> {
    const res = await this.http.request<AuthResult>(
      "POST",
      "/v1/account/sign-up",
      {
        auth: "none",
        body: {
          project_id: this.http.getProjectId(),
          email: input.email,
          password: input.password,
          name: input.name,
        },
      }
    );
    // MFA 分支：无 tokens，先返回 challenge 信息，调用方应引导二次认证。
    if (res.mfa_required) return res;
    this.http.setAccessToken(res.tokens?.access_token);
    return res;
  }

  async signIn(input: {
    email: string;
    password: string;
  }): Promise<AuthResult> {
    const res = await this.http.request<AuthResult>(
      "POST",
      "/v1/account/sign-in",
      {
        auth: "none",
        body: {
          project_id: this.http.getProjectId(),
          email: input.email,
          password: input.password,
        },
      }
    );
    if (res.mfa_required) return res;
    this.http.setAccessToken(res.tokens?.access_token);
    return res;
  }

  async signOut(): Promise<void> {
    await this.http.request<void>("POST", "/v1/account/sign-out", {
      body: { project_id: this.http.getProjectId() },
    });
    this.http.setAccessToken(undefined);
  }

  async refresh(refreshToken: string): Promise<TokenBundle> {
    const res = await this.http.request<{ tokens: TokenBundle }>("POST", "/v1/account/refresh", {
      auth: "none",
      body: {
        project_id: this.http.getProjectId(),
        refresh_token: refreshToken,
      },
    });
    this.http.setAccessToken(res.tokens.access_token);
    return res.tokens;
  }

  async me(): Promise<Account> {
    return this.http.request<Account>("GET", "/v1/account/me", {
      query: { project_id: this.http.getProjectId() },
    });
  }

  async updateAccount(input: {
    name?: string;
    email?: string;
    password?: string;
    old_password?: string;
    // 改邮箱时必填：新邮箱验证链接模板（staging：验证通过前 email 保持旧值）。
    url?: string;
  }): Promise<Account> {
    return this.http.request<Account>("PATCH", "/v1/account", { body: input });
  }

  // 消费邮件链接中的一次性 secret 完成邮箱变更（公开方法，无需登录：
  // 凭 user_id + secret，与 recovery 同一安全模型——随机 secret + TTL + 一次性消费）。
  async confirmEmailChange(input: {
    user_id: string;
    secret: string;
  }): Promise<Account> {
    return this.http.request<Account>("PUT", "/v1/account/email-change", {
      auth: "none",
      body: {
        project_id: this.http.getProjectId(),
        user_id: input.user_id,
        secret: input.secret,
      },
    });
  }

  async listSessions(): Promise<Session[]> {
    const res = await this.http.request<{ sessions: Session[] }>("GET", "/v1/account/sessions");
    return res.sessions ?? [];
  }

  async deleteSession(sessionId: string): Promise<void> {
    await this.http.request<void>("DELETE", `/v1/account/sessions/${sessionId}`);
  }

  async deleteSessions(keepCurrent = false): Promise<void> {
    // keep_current 经 grpc-gateway 绑定为查询参数（DELETE 无 body）。
    await this.http.request<void>("DELETE", "/v1/account/sessions", {
      query: { keep_current: String(keepCurrent) },
    });
  }

  async getPrefs(): Promise<Record<string, unknown>> {
    const res = await this.http.request<{ prefs: Record<string, unknown> }>(
      "GET",
      "/v1/account/prefs"
    );
    return res.prefs ?? {};
  }

  async updatePrefs(prefs: Record<string, unknown>): Promise<Record<string, unknown>> {
    const res = await this.http.request<{ prefs: Record<string, unknown> }>(
      "PUT",
      "/v1/account/prefs",
      { body: { prefs } }
    );
    return res.prefs ?? {};
  }

  async createOAuth2Session(input: {
    provider: string;
    success: string;
    failure: string;
  }): Promise<{ redirect_url: string }> {
    return this.http.request("GET", `/v1/account/sessions/oauth2/${encodeURIComponent(input.provider)}`, {
      auth: "none",
      query: {
        project_id: this.http.getProjectId(),
        success: input.success,
        failure: input.failure,
      },
    });
  }

  async createOAuth2TokenSession(input: {
    provider: string;
    code: string;
    state: string;
    success?: string;
    failure?: string;
  }): Promise<{ account: Account; tokens: TokenBundle }> {
    const res = await this.http.request<{ account: Account; tokens: TokenBundle }>(
      "POST",
      `/v1/account/sessions/oauth2/${encodeURIComponent(input.provider)}/token`,
      {
        auth: "none",
        body: {
          project_id: this.http.getProjectId(),
          code: input.code,
          state: input.state,
          success: input.success,
          failure: input.failure,
        },
      }
    );
    this.http.setAccessToken(res.tokens.access_token);
    return res;
  }

  async createEmailOTP(input: {
    email: string;
  }): Promise<{ challenge_id: string; expire_at: string }> {
    return this.http.request("POST", "/v1/account/sessions/email-otp", {
      auth: "none",
      body: {
        project_id: this.http.getProjectId(),
        email: input.email,
      },
    });
  }

  async createEmailOTPSession(input: {
    email: string;
    challenge_id: string;
    otp: string;
  }): Promise<{ account: Account; tokens: TokenBundle }> {
    const res = await this.http.request<{ account: Account; tokens: TokenBundle }>(
      "POST",
      "/v1/account/sessions/email-otp/verify",
      {
        auth: "none",
        body: {
          project_id: this.http.getProjectId(),
          email: input.email,
          challenge_id: input.challenge_id,
          otp: input.otp,
        },
      }
    );
    this.http.setAccessToken(res.tokens.access_token);
    return res;
  }

  async createPhoneOTP(input: {
    phone: string;
  }): Promise<{ challenge_id: string; expire_at: string }> {
    return this.http.request("POST", "/v1/account/sessions/phone-otp", {
      auth: "none",
      body: {
        project_id: this.http.getProjectId(),
        phone: input.phone,
      },
    });
  }

  async createPhoneOTPSession(input: {
    phone: string;
    challenge_id: string;
    otp: string;
  }): Promise<{ account: Account; tokens: TokenBundle }> {
    const res = await this.http.request<{ account: Account; tokens: TokenBundle }>(
      "POST",
      "/v1/account/sessions/phone-otp/verify",
      {
        auth: "none",
        body: {
          project_id: this.http.getProjectId(),
          phone: input.phone,
          challenge_id: input.challenge_id,
          otp: input.otp,
        },
      }
    );
    this.http.setAccessToken(res.tokens.access_token);
    return res;
  }

  async createWeChatMiniProgramSession(input: {
    code: string;
  }): Promise<{ account: Account; tokens: TokenBundle }> {
    const res = await this.http.request<{ account: Account; tokens: TokenBundle }>(
      "POST",
      "/v1/account/sessions/wechat/miniprogram",
      {
        auth: "none",
        body: {
          project_id: this.http.getProjectId(),
          code: input.code,
        },
      }
    );
    this.http.setAccessToken(res.tokens.access_token);
    return res;
  }

  // 创建匿名会话；返回的匿名用户可通过 signIn 升级为正式账号。
  async createAnonymousSession(): Promise<{ account: Account; tokens: TokenBundle }> {
    const res = await this.http.request<{ account: Account; tokens: TokenBundle }>(
      "POST",
      "/v1/account/sessions/anonymous",
      {
        auth: "none",
        body: { project_id: this.http.getProjectId() },
      }
    );
    this.http.setAccessToken(res.tokens.access_token);
    return res;
  }

  // 生成第三方账号绑定授权链接（需已登录）。
  async createOAuth2LinkSession(input: {
    provider: string;
    success: string;
    failure: string;
  }): Promise<{ redirect_url: string }> {
    return this.http.request("GET", `/v1/account/sessions/oauth2/${encodeURIComponent(input.provider)}/link`, {
      query: {
        project_id: this.http.getProjectId(),
        success: input.success,
        failure: input.failure,
      },
    });
  }

  // 用 OAuth 回调 code 绑定第三方账号（需已登录）。
  async createOAuth2LinkTokenSession(input: {
    provider: string;
    code: string;
    state: string;
  }): Promise<Account> {
    return this.http.request<Account>(
      "POST",
      `/v1/account/sessions/oauth2/${encodeURIComponent(input.provider)}/link/token`,
      {
        body: {
          project_id: this.http.getProjectId(),
          code: input.code,
          state: input.state,
        },
      }
    );
  }

  // 发送邮箱验证邮件（url 为带 {{code}} 占位的确认链接模板）。
  async createVerification(input: {
    url: string;
  }): Promise<{ user_id: string; expire_at: string }> {
    return this.http.request("POST", "/v1/account/verification", {
      body: {
        project_id: this.http.getProjectId(),
        url: input.url,
      },
    });
  }

  // 用邮件中的 secret 确认邮箱（公开：无需登录）。
  async updateVerification(input: {
    user_id: string;
    secret: string;
  }): Promise<Account> {
    return this.http.request<Account>("PUT", "/v1/account/verification", {
      auth: "none",
      body: {
        project_id: this.http.getProjectId(),
        user_id: input.user_id,
        secret: input.secret,
      },
    });
  }

  // 发送密码找回邮件。
  async createRecovery(input: {
    email: string;
    url: string;
  }): Promise<void> {
    await this.http.request<void>("POST", "/v1/account/recovery", {
      auth: "none",
      body: {
        project_id: this.http.getProjectId(),
        email: input.email,
        url: input.url,
      },
    });
  }

  // 用邮件中的 secret 重置密码（公开：无需登录）。
  async updateRecovery(input: {
    user_id: string;
    secret: string;
    password: string;
  }): Promise<void> {
    await this.http.request<void>("PUT", "/v1/account/recovery", {
      auth: "none",
      body: {
        project_id: this.http.getProjectId(),
        user_id: input.user_id,
        secret: input.secret,
        password: input.password,
      },
    });
  }

  // ---- MFA ----
  async listFactors(): Promise<Factor[]> {
    const res = await this.http.request<{ factors: Factor[] }>("GET", "/v1/account/mfa");
    return res.factors ?? [];
  }

  // 创建 TOTP 因子；secret/otpauth_url 仅本次返回明文，之后不回显。
  async createTOTPFactor(): Promise<TOTPFactor> {
    return this.http.request<TOTPFactor>("POST", "/v1/account/mfa/totp", {
      body: {},
    });
  }

  // 校验并激活 TOTP 因子。
  async verifyTOTPFactor(input: {
    factor_id: string;
    code: string;
  }): Promise<Factor> {
    return this.http.request<Factor>("PUT", "/v1/account/mfa/totp", {
      body: input,
    });
  }

  // 删除 MFA 因子：pending 因子直接删除；verified 因子必须携带 TOTP code
  // 二次验证（code 经 query 传递：DELETE ?code=...，R05-P1-4）。
  async deleteFactor(factorId: string, code?: string): Promise<void> {
    await this.http.request<void>("DELETE", `/v1/account/mfa/${factorId}`, {
      query: code ? { code } : undefined,
    });
  }

  // 登录二次验证：携带 signIn/signUp 返回的 challenge_token 完成挑战，
  // 成功后自动保存 token。
  async createMFASession(input: {
    challenge_token: string;
    factor_id: string;
    code: string;
  }): Promise<{ account: Account; tokens: TokenBundle }> {
    const res = await this.http.request<{ account: Account; tokens: TokenBundle }>(
      "POST",
      "/v1/account/mfa/challenge",
      {
        auth: "none",
        body: {
          project_id: this.http.getProjectId(),
          challenge_token: input.challenge_token,
          factor_id: input.factor_id,
          code: input.code,
        },
      }
    );
    this.http.setAccessToken(res.tokens.access_token);
    return res;
  }

  // 用当前会话换取一次性 JWT（用于服务端安全回调/Webhook）。
  async createJWT(): Promise<{ token: string }> {
    return this.http.request("POST", "/v1/account/jwt", { body: {} });
  }

  // ---- Magic URL ----
  async createMagicURLSession(input: {
    email: string;
    url: string;
  }): Promise<{ challenge_id: string; expire_at: string }> {
    return this.http.request("POST", "/v1/account/sessions/magic-url", {
      auth: "none",
      body: {
        project_id: this.http.getProjectId(),
        email: input.email,
        url: input.url,
      },
    });
  }

  // 用邮件中的 secret 完成 Magic URL 登录。
  async updateMagicURLSession(input: {
    user_id: string;
    secret: string;
  }): Promise<{ account: Account; tokens: TokenBundle }> {
    const res = await this.http.request<{ account: Account; tokens: TokenBundle }>(
      "PUT",
      "/v1/account/sessions/magic-url",
      {
        auth: "none",
        body: {
          project_id: this.http.getProjectId(),
          user_id: input.user_id,
          secret: input.secret,
        },
      }
    );
    this.http.setAccessToken(res.tokens.access_token);
    return res;
  }

  // 账号操作日志（最近 limit 条，默认 25）。
  async listLogs(limit?: number): Promise<LogEntry[]> {
    const res = await this.http.request<{ logs: LogEntry[] }>("GET", "/v1/account/logs", {
      query: { limit: limit ?? 25 },
    });
    return res.logs ?? [];
  }
}
